SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -0,0 +1,144 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSProfilesManager.h"
+#import "OWSMessageSender.h"
+#import "SecurityUtils.h"
+#import "TSStorageManager.h"
+#import "TextSecureKitEnv.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString *const kOWSProfilesManager_Collection = @"kOWSProfilesManager_Collection";
+// This key is used to persist the local user's profile key.
+NSString *const kOWSProfilesManager_LocalProfileKey = @"kOWSProfilesManager_LocalProfileKey";
+
+// TODO:
+static const NSInteger kProfileKeyLength = 16;
+
+@interface OWSProfilesManager ()
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+@property (nonatomic, readonly) OWSMessageSender *messageSender;
+
+@property (nonatomic, readonly, nullable) NSData *localProfileKey;
+
+@end
+
+#pragma mark -
+
+@implementation OWSProfilesManager
+
+@synthesize localProfileKey = _localProfileKey;
+
++ (instancetype)sharedManager
+{
+    static OWSProfilesManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    OWSMessageSender *messageSender = [TextSecureKitEnv sharedEnv].messageSender;
+
+    return [self initWithStorageManager:storageManager messageSender:messageSender];
+}
+
+- (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
+                         messageSender:(OWSMessageSender *)messageSender
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(storageManager);
+    OWSAssert(messageSender);
+
+    _storageManager = storageManager;
+    _messageSender = messageSender;
+
+    OWSSingletonAssert();
+
+    // Register this manager with the message sender.
+    // This is a circular dependency.
+    [messageSender setProfilesManager:self];
+
+    // Try to load.
+    _localProfileKey = [self.storageManager objectForKey:kOWSProfilesManager_LocalProfileKey
```

Comment:
The key-value accessor methods in TSStorageManager are really useful.  I propose that we move them to a category on `YapDatabaseReadTransaction` and `YapDatabaseReadWriteTransaction`.  

I'm not sure that I support eliminating `TSStorageManager.dbReadConnection` and `TSStorageManager.dbReadWriteConnection` entirely, without reviewing all of the places that they're used. 

I'm making cards "below the line"?